### PR TITLE
Configure legacy membership instead of resolving during fallback

### PR DESCRIPTION
**Pull request by jdom**

### CODE REVIEW

File: src/OrleansConsulUtils/LegacyConsulMembershipConfigurator.cs [hunk 1]
```diff
@@ -1,34 +1,15 @@
...
 
 namespace Orleans.Runtime.MembershipService
 {
     /// <inheritdoc cref="ILegacyMembershipConfigurator"/>
     public class LegacyConsulMembershipConfigurator : ILegacyMembershipConfigurator
     {
-        private readonly ILoggerFactory loggerFactory;
-        private readonly GlobalConfiguration globalConfiguration;
-        public LegacyConsulMembershipConfigurator(ILoggerFactory loggerFactory, GlobalConfiguration globalConfiguration)
+        public void ConfigureServices(GlobalConfiguration configuration, IServiceCollection services)
```
Comment (xiazen) [inline src/OrleansConsulUtils/LegacyConsulMembershipConfigurator.cs:10]: as a design principle. we shouldn't put dependency in APIs. As in this API is not extensible at all. What if there's one more dependency, which we need to inject from DI anyways...

File: src/OrleansRuntime/Hosting/CoreHostingExtensions.cs [hunk 2]
```diff
@@ -68,11 +67,17 @@ public static ISiloHostBuilder ConfigureLocalHostPrimarySilo(this ISiloHostBuild
...
         public static ISiloHostBuilder UseGrainBasedMembership(this ISiloHostBuilder builder)
         {
-            return builder.ConfigureServices(services => services.TryAddSingleton<UseGrainBasedMembershipFlag>());
+            return builder.ConfigureServices(services => services.UseGrainBasedMembership());
+        }
+
+        /// <summary>
+        /// Configure silo to use Grain-based membership
+        /// </summary>
+        public static IServiceCollection UseGrainBasedMembership(this IServiceCollection services)
```
Comment (xiazen) [inline src/OrleansRuntime/Hosting/CoreHostingExtensions.cs]: remove this.. I added it above

File: src/OrleansConsulUtils/LegacyConsulMembershipConfigurator.cs (hunk 1 above)
Comment (jdom) [reply src/OrleansConsulUtils/LegacyConsulMembershipConfigurator.cs:10]: Hmm, not sure I follow, this IS just a way to turn GlobalConfiguration into whatever registration we need in DI... it's not injecting general dependencies, it's just passing the only relevant piece of info. See that I removed logger or any other potential dependencies, this should just be a conversion from legacy config to configuring DI

Comment (xiazen) [reply src/OrleansConsulUtils/LegacyConsulMembershipConfigurator.cs:10]: If you look from the lens of ILegacyMembershipConfigurator, gobalConfiguration and ServiceCollection is dependency for the configurator. 
I know it is not the case, here, but what if the configurator needs a third dependency from DI? Then this interface is not extensible.

Comment (jdom) [reply src/OrleansConsulUtils/LegacyConsulMembershipConfigurator.cs:10]: The point is that this doesn't run after the silo started. This runs when the host builder is configuring things. So it uses the same extension methods that a normal user would use for configuring the host. If it needs something from the container, it means that the user wouldn't be able to configure the same membership by not using the legacy converter.

**Comment by jdom** [comment]
Pushed a commit that configures both the client and the silo from the legacy configuration before the container is built.
This wouldn't require the legacy config objects to be registered in the container once we remove all the usages.
Basically it performs all the registrations in the container as soon as you pass the legacy config object, and users can override the configuration afterwards.

/cc @ReubenBond @attilah

File: src/OrleansConsulUtils/LegacyConsulMembershipConfigurator.cs (hunk 1 above)
Comment (xiazen) [reply src/OrleansConsulUtils/LegacyConsulMembershipConfigurator.cs:10]: > If it needs something from the container, it means that the user wouldn't be able to configure the same membership by not using the legacy converter.

I'm not sure what you mean by this. Our code behavior are very similar, we both configure membership when host builder is configuring things. 

> it uses the same extension methods that a normal user would use for configuring the host.

Sure this is better. My way is more preserving its original logic, which is what we talked about and agreed. But your way is better. I will take it then,  I will take a look on your second commit and see if we can merge this.

File: src/OrleansRuntime/MembershipService/GrainBasedMembershipTable.cs [hunk 3]
```diff
@@ -1,28 +1,113 @@
...
 {
-    //empty marker class inject into DI to see if user configure Silo to use MembershiptableGrain
-    internal class UseGrainBasedMembershipFlag
+    internal class GrainBasedMembershipTable : IMembershipTable
     {
+        private readonly IServiceProvider serviceProvider;
+        private readonly Logger logger;
+        private IMembershipTableGrain grain;
+
+        public GrainBasedMembershipTable(IServiceProvider serviceProvider, LoggerWrapper<MembershipTableFactory> logger)
```
Comment (xiazen) [inline src/OrleansRuntime/MembershipService/GrainBasedMembershipTable.cs:18]: Why are we using LoggerWrapper instead of ILogger?

Comment (jdom) [reply src/OrleansRuntime/MembershipService/GrainBasedMembershipTable.cs:18]: I just moved the code that was using `Logger` from `MembershipTableFactory` without unnecessary changes to keep the diff to a minimum (as it's already hard to follow when moving methods from file to file). I can certainly do the switch to `ILogger` if you already diffed and reviewed the code, if there's value of doing it now for this occurrence before the big switch to `ILogger`. Just let me know what to do.

File: src/OrleansRuntime/Silo/Silo.cs [hunk 4]
```diff
@@ -178,6 +178,7 @@ internal Silo(SiloInitializationParameters initializationParams, IServiceProvide
                 var serviceCollection = new ServiceCollection();
                 serviceCollection.AddSingleton<Silo>(this);
                 serviceCollection.AddSingleton(initializationParams);
+                serviceCollection.AddLegacyClusterConfigurationSupport(config);
```
Comment (xiazen) [inline src/OrleansRuntime/Silo/Silo.cs:181]: Isn't this should be in SiloHostBuilder.UseConfiguration method ? Why is it here?

File: src/OrleansRuntime/MembershipService/GrainBasedMembershipTable.cs (hunk 3 above)
Comment (xiazen) [reply src/OrleansRuntime/MembershipService/GrainBasedMembershipTable.cs:18]: sure no strong objection on this one .

File: src/OrleansRuntime/Silo/Silo.cs (hunk 4 above)
Comment (jdom) [reply src/OrleansRuntime/Silo/Silo.cs:181]: Because this is the path where the user doesn't use `SiloBuilder`, and instead instantiates `Silo` directly (or via `AzureSilo`, or `SiloHost`). So this path effectively ONLY supports legacy configuration. This path will probably go away very soon, @ReubenBond was bringing it up as something to get rid of for 2.0, and I think I agree, but of course need to be dealt separately as just dropping support in this PR should not be by itself a breaking change. We are tracking it for discussion for post 2.0-beta

Comment (xiazen) [reply src/OrleansRuntime/Silo/Silo.cs:181]: ah nvm . you added it there
